_posts/2019-07-10-feature-columns/feature_columns.Rmd

```diff
@@ -0,0 +1,566 @@
+
+Nice as this sounds, why don't we see it used all the time? Well, creating a Keras network that processes a mix of numeric and categorical data used to require a bit of an effort. With TensorFlow's new _feature columns_, usable from R through a combination of `tfdatasets` and `keras`, there is a much easier way to achieve this. What's more, `tfdatasets` follows the popular [recipes](https://cran.r-project.org/web/packages/recipes/index.html) idiom to initialize, refine and apply a feature specification `%>%`-style. And finally, there are ready-made steps for bucketizing a numeric column, or hashing it, or creating _crossed columns_ to capture interactions.
+
+This post introduces feature specs starting from a scenario where they don't exist: basically, the status quo until very recently. Imagine you have a dataset like that from the [Porto Seguro car insurance competition](https://www.kaggle.com/c/porto-seguro-safe-driver-prediction) where some of the columns are numeric, and some are categorical. You want to train a fully connected network on it, with all categorical columns fed into embedding layers. How can you do that? We then contrast this with the feature spec way, which makes things _a lot_ easier -- especially when there's a lot of categorical columns.
+In a second applied example, we demonstrate the use of _crossed columns_ on the _rugged_ dataset from Richard McElreath's _rethinking_ package. Here, we also direct attention to a few technical details that are worth knowing about. 
+
+## Mixing numeric data and embeddings, the pre-feature-spec way
+
+Our first example dataset is taken from Kaggle. Two years ago, Brazilian car insurance company Porto Seguro asked participants to predict [how likely it is a car owner will file a claim](https://www.kaggle.com/c/porto-seguro-safe-driver-prediction) based on a mix of characteristics collected during the previous year. The dataset is comparatively large -- there are ~ 600,000 rows in the training set, with 57 predictors. Among others, features are named so as to indicate the type of the data -- binary, categorical, or continuous/ordinal.
+While it's common in competitions to try to reverse-engineer column meanings, here we just make use of the type of the data, and see how far that gets us.
+
+Concretely, this means we want to
+
+- use binary features just the way they are, as zeroes and ones,
+- scale the remaining numeric features to mean 0 and variance 1, and
+- embed the categorical variables (each one by itself).
+
+We'll then define a dense network to predict `target`, the binary outcome. So first, let's see how we could get our data into shape, as well as build up the network, in a "manual", pre-feature-columns way.
+
+When loading libraries, we already use the versions we'll need very soon: Tensorflow 2 (>= beta 1), and the development (= Github) versions of `tfdatasets` and `keras`:
```

Comment:
> When loading libraries, we already use the versions we'll need very soon:

By "very soon" do you mean in this post? If so, I'd simplify this sentence to (see suggestion edit):